cloudinit/config/cc_landscape.py

```diff
@@ -151,8 +151,17 @@ def handle(name: str, cfg: Config, cloud: Cloud, args: list) -> None:
             ]
         )
     )
-    subp.subp(["landscape-config", "--silent"] + cmd_params)
-    util.write_file(LS_DEFAULT_FILE, "RUN=1\n")
+    try:
+        subp.subp(["landscape-config", "--silent", "--is-registered"], rcs=[5])
```

Comment:
This will not raise a ProcessExecutionError if you pass rcs=[5] on a registered system.
```
# Note I'm registered in this instance and the exit 5
root@nnn:~# landscape-config --silent --is-registered

Registered:    False
Config Path:   /etc/landscape/client.conf
Data Path      /var/lib/landscape/client
root@nnn:~# echo $?
5
# Note that subp with rcs=[5] allows the subprocess to receive the exit code 5 without raising a ProcessExecutionError

root@nnn:~# python3 -c 'from cloudinit.subp import subp; subp(["landscape-config", "--silent", "--is-registered"], rcs=[5])'

# If we drop the rcs=[5], we get the desired ProcessExecutionError

root@nnn:~# python3 -c 'from cloudinit.subp import subp; subp(["landscape-config", "--silent", "--is-registered"])'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python3/dist-packages/cloudinit/subp.py", line 277, in subp
    raise ProcessExecutionError(
cloudinit.subp.ProcessExecutionError: Unexpected error while running command.
Command: ['landscape-config', '--silent', '--is-registered']
Exit code: 5
Reason: -
Stdout: 
        Registered:    False
        Config Path:   /etc/landscape/client.conf
        Data Path      /var/lib/landscape/client
Stderr: 
```
So,  I think what you want is actually the following and instead of checking for exit_code == 0 below check for exit_code == 5
```suggestion
        subp.subp(["landscape-config", "--silent", "--is-registered"])
```

Reply:
landscape-client code states that exit code **5** stands for [NOT REGISTERED](https://github.com/canonical/landscape-client/blob/51b8d3243278d1d5405db832c3471d04d784de13/landscape/client/configuration.py#L40) and 0 if client is already registered. Based on this facts,  i think we should continue (on exit code 5) and register the client and log out a warning when exit code is 0. 
I made following tests on Focal with landscape-client 23.02:

```
# Already registered to Landscape
root@test01:/# landscape-config --silent --is-registered

Registered:    True
Config Path:   /etc/landscape/client.conf
Data Path      /var/lib/landscape/client
Account Name:  standalone
root@test01:/# echo $?
0

# Not registered to Landscape
root@test01:/# landscape-config --silent --is-registered

Registered:    False
Config Path:   /etc/landscape/client.conf
Data Path      /var/lib/landscape/client
root@test01:/# echo $?
5
```

I am aware that an exception will be thrown for 
`subp.subp(["landscape-config", "--silent", "--is-registered"], rcs=[5])`. But we do not log the exception to file, which could possibly irritate a user, rather than writing a simple warning message that the client is already registered.


What do you think?

Reply:
oops. You are correct @chifac08. I misinterpreted the fact that your rcs=[5] lacked the exit 0, which would end up raising ProcessExecutionError on exit 0 in the case when already registered. Ok this makes perfect sense. And logging a warning is good nowadays too as newer versions of cloud-init will exit(2) when warnings are present, making that conditon more conspicious. I also think we don't want to log full exceptions or tracebacks with warning logs can suffice to explain the situation clearly enough. So, your approach works fine here.

Reply:
Ok, resolving this message as your approach will work fine as long as --is-registered parameter is available from lansdcape-config command. (Again we can wait for SRU of landscape-client for this feature on all supported cloud-init Ubuntu series.